File: web/src/apis/types/structs.ts
import * as enums from "./enums";

export interface Class{
    name: string
}
export interface SingleClass{
    id: number,
    name: string
}
export interface SingleUserWithoutAuth{
    id: number,
    name: string
}
export interface SingleUser extends SingleUserWithoutAuth{
    auth: number
}
export type ListClassesResponse = Array<SingleClass>
export interface ClassInfoResponse{
    name: string,
    students: Array<SingleUser>,
    teachers: Array<SingleUser>
}
export interface ClassStudentNum{
    num: number
}
export interface UserLoginResponse{
    token: string,
    id: number
}
export type SearchUsersResponse = Array<SingleUser>
export interface UserBasicInfoResponse{
    clsName: string,
    userName: string
}
export interface UserInfoResponse{
    name: string,
    school_id: number,
    cls: number,
    auth: number,
    clsName: string
}
export interface PictureResponse{
    hash: string,
    type: string
}
export interface ThoughtInfoResponse{
    status: enums.ThoughtStatus,
    thought: string,
    pics: Array<PictureResponse>,
    reward: number,
    everRepulsed: boolean,
    reason: string
}
export interface StudentThoughtsResponse{
    accepted: Array<ThoughtInfoResponse>,
    unsubmitted: Array<ThoughtInfoResponse>,
    draft: Array<ThoughtInfoResponse>,
    unaudited: Array<ThoughtInfoResponse>
}
export interface StudentStatResponse{
    inside: number,
    outside: number,
    large: number
}
export interface SingleNotice{
    id: number,
    title: string,
    content: string,
    sender: number,
    sendtime: string,
    deadtime: string,
    senderName: string
}
export type SearchNoticesResponse = Array<SingleNotice>
export interface SingleSignup{
    volId: number,
    volName: string,
    stuId: number,
    stuName: string
}
export type ListSignupResponse = Array<SingleSignup>
export interface SingleVolunteer{
    id: number,
    name: string,
    time: string,
    status: number,
    signable: boolean,
    joiners: Array<SingleUserWithoutAuth>,
    holderName: string
}
export type SearchVolunteersResponse = Array<SingleVolunteer>
export interface SearchNotices{
    sender?: number,
    receiver?: number,
    cls?: number,
    school?: number
}
export interface NoticeBody{
    title: string,
    content: string,
    deadtime: string
}
export interface Notice extends NoticeBody{
    targets: Array<number>,
    anonymous: boolean
}
export interface SchoolNotice extends NoticeBody{
    anonymous: boolean
}
export interface Report{
    report: string
}
export interface SingleReport{
    content: string,
    reporter: number,
    reporterName: string,
    time: string
}
export type FetchReportsResponse = Array<SingleReport>
export interface Signup{
    students: Array<number>
}
export interface SearchThoughts{
    cls?: number,
    status?: enums.ThoughtStatus,
    student?: number,
    volunteer?: number
}
export interface SearchStudentThoughts{
    status?: enums.ThoughtStatus
}
export interface SingleThought{
    volId: number,
    stuId: number,
    status: enums.ThoughtStatus,
    stuName: string,
    volName: string,
    volTime: string
}
export type SearchThoughtsResponse = Array<SingleThought>
export interface BasePictrure{
    type: string
}
export interface ExistedPicture extends BasePictrure{
    hash: string,
    type: string
}
export interface Base64Picture extends BasePictrure{
    base64: string,
    type: string
}
export type Picture = ExistedPicture | Base64Picture
export interface Thought{
    thought: string,
    pictures: Array<ExistedPicture | Base64Picture>
}
export interface Login{
    id: string,
    pwd: string,
    devideId: string
}
export interface SearchUsers{
    name?: string,
    cls?: number,
    auth?: number
}
export interface ModPwd{
    old: string,
    neo: string
}
export interface User{
    name: string,
    cls: number,
    auth: number
}
export interface OneUser extends User{
    id: number
}
export interface Users{
    users: Array<OneUser>
}
export interface ClassVol{
    id: number,
    max: number
}
export interface ClassVolWithName extends ClassVol{
    name: string
}
export interface ListVolunteers{
    cls?: number
}
export interface SearchVolunteers{
    holder?: number,
    student?: number,
    cls?: number,
    name?: string,
    status?: enums.VolStatus,
    signable?: boolean
}
export interface VolunteerBody{
    name: string,
    description: string,
    time: string,
    type: enums.VolType,
    reward: number
}
export interface Volunteer extends VolunteerBody{
    classes: Array<ClassVol>
}
export interface VolunteerInfoResponse{
    name: string,
    description: string,
    time: string,
    status: enums.VolStatus,
    type: enums.VolType,
    reward: number,
    signable: boolean,
    classes: Array<ClassVolWithName>,
    joiners: Array<SingleUserWithoutAuth>,
    holder: number,
    holderName: string
}
export interface AppointedVolunteer extends VolunteerBody{
    joiners: Array<number>
}
export interface Repulse{
    reason: string
}
export interface Accept{
    reward: number
}
export interface FetchPicture{
    url: string
}
export interface PublicNoticeNotNone{
    title: string,
    content: string
}
export type PublicNotice = PublicNoticeNotNone | null
export interface ModOthersPwd{
    pwd: string
}
export interface SpecialVolunteer{
    name: string,
    type: enums.VolType,
    reward: number,
    joiners: Array<number>
}

File: backend/zvms/typing/structs.py
from zvms.typing.checker import *
from zvms.res import VolType, VolStatus, ThoughtStatus

class Class(Object):
    name = Len(String, 1, 6)

class SingleClass(Object):
    id = Int
    name = String

class SingleUserWithoutAuth(Object):
    id = Int
    name = String

class SingleUser(SingleUserWithoutAuth):
    auth = Int

ListClassesResponse = Array(SingleClass())
    
class ClassInfoResponse(Object):
    name = String
    students=Array(SingleUser())
    teachers=Array(SingleUser())

class UserLoginResponse(Object):
    token = String
    id = Int

SearchUsersResponse = Array(SingleUser())

class UserInfoResponse(Object):
    name = String
    cls = Int
    auth = Int
    clsName = String

class PictureResponse(Object):
    hash = String
    type = String

class ThoughtInfoResponse(Object):
    status = Enum(ThoughtStatus)
    reason = String
    thought = String
    reward = Int
    pics = Array(PictureResponse())

class StudentThoughtsResponse(Object):
    accepted = Array(ThoughtInfoResponse())
    unsubmitted = Array(ThoughtInfoResponse())
    draft = Array(ThoughtInfoResponse())
    unaudited = Array(ThoughtInfoResponse())

class StudentStatResponse(Object):
    inside = Int
    outside = Int
    large = Int

class SingleNotice(Object):
    id = Int
    title = String
    content = String
    sender = Int
    sendtime = DateTime
    deadtime = DateTime
    senderName = String
    
SearchNoticesResponse = Array(SingleNotice())

class SingleSignup(Object):
    volId = Int
    volName = String
    stuId = Int
    stuName = String
    
ListSignupResponse = Array(SingleSignup())

class SingleVolunteer(Object):
    id = Int
    name = String
    time = String
    status = Int
    signable = Boolean
    joiners = Array(SingleUserWithoutAuth())
    holderName = String
    
SearchVolunteersResponse = Array(SingleVolunteer())

class VolunteerInfoResponse(Object):
    name = String
    description = String
    time = String
    status = Enum(VolStatus)
    type = Enum(VolType)
    reward = Int
    signable = Boolean
    joiners = Array(SingleUserWithoutAuth())
    holder = Int
    holderName = String

class SearchNotices(Optional):
    sender = Parsable(Int)
    receiver = Parsable(Int)
    cls = Parsable(Int)
    school = Parsable(Int)

class NoticeBody(Object):
    title = Len(String, 1, 33)
    content = Len(String, 1, 1025)
    deadtime = DateTime
    
class Notice(NoticeBody):
    targets = Array(Int)

class SchoolNotice(NoticeBody):
    anonymous = Boolean

class Report(Object):
    report = Len(String, 1, 256)

class SingleReport(Object):
    content = String
    reporter = Int
    reporterName = String
    time = String

FetchReportsResponse = Array(SingleReport())
    
class Signup(Object):
    students = Array(Int)
    
class SearchThoughts(Optional):
    cls = Parsable(Int)
    status = ParsableEnum(ThoughtStatus)
    student = Parsable(Int)
    volunteer = Parsable(Int)

class SingleThought(Object):
    volId = Int
    stuId = Int
    status = Enum(ThoughtStatus)
    stuName = String
    volName = String

SearchThoughtsResponse = Array(SingleThought())

class ExistedPicture(Object):
    hash = String
    type = Len(String, 3, 6)

class Base64Picture(Object):
    base64 = String
    type = Len(String, 3, 6)

Picture = Union(ExistedPicture(), Base64Picture())

class Thought(Object):
    thought = Len(String, 0, 1025)
    pictures = Array(Picture)

class Login(Object):
    id = String
    pwd = Len(String, 32, 33)

class SearchUsers(Optional):
    name = String
    cls = Parsable(Int)
    auth = Parsable(Int)
    
class ModPwd(Object):
    old = Len(String, 32, 33)
    neo = Len(String, 32, 33)

class User(Object):
    name = Len(String, 1, 6)
    cls = Int
    auth = Int

class OneUser(User):
    id = Int

class Users(Object):
    users = Array(OneUser())

class ClassVol(Object):
    id = Int
    max = Int
    
class SearchVolunteers(Optional):
    holder = Parsable(Int)
    student = Parsable(Int)
    cls = Parsable(Int)
    name = String
    status = ParsableEnum(VolStatus)
    signable = Parsable(Boolean)
    
class VolunteerBody(Object):
    name = Len(String, 1, 33)
    description = Len(String, 1, 1025)
    time = DateTime
    type = Enum(VolType)
    reward = Int

class Volunteer(VolunteerBody):
    classes = Array(ClassVol())

class AppointedVolunteer(VolunteerBody):
    joiners = Array(Int)

class Repulse(Object):
    reason = Len(String, 0, 65)

class Accept(Object):
    reward = Int

class FetchPicture(Object):
    url = String

class PublicNoticeNotNone(Object):
    title = String
    content = String

PublicNotice =  Nullable(PublicNoticeNotNone())